### ISSUE TITLE
[$2000] Workspace - Deleted user from workspace can chat in previously created workspace room

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#15744]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create WS
2. Add member X
3. Create room (member X is added, as he is in WS)
4. Send some message s in the room
5. Delete member X from WS
6. Send some messages in the room

## Expected Result:
Member X should not see messages in the room as was eliminated from workspace

## Actual Result:
Member X continues to stay in the room and can see and send messages, but with room tag "unavailable workspace"

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.80.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/223731729-2947b5bc-c0d8-4180-ab65-eb5321055971.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019fa2485ed55ee079</li>
        <li>Upwork Job ID: 1638546363408375808</li>
        <li>Last Price Increase: 2023-03-29</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kevinksullivan (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by kevinksullivan** [comment]
Hi @kbecciv was there a slack channel for this one?

**Comment by kevinksullivan** [comment]
Hm I'm also unable to reproduce this. What I did was this:

1. Create WS on one browser
2. Add member X
3. Go to incognito and sign is as member X, open the announce room
4. Go to announce room for WS, write message from the creator
5. Confirm in incognito that member X sees message
7. Delete member X from WS
8. Send some messages in the room

member X was immediately removed 

<img width="961" alt="image" src="https://user-images.githubusercontent.com/22508290/224230315-74599381-7605-4c6d-b617-cddc59b90fb7.png">

**Comment by MelvinBot** [comment]
@kevinksullivan Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@kevinksullivan Eep! 4 days overdue now. Issues have feelings too...

**Comment by MelvinBot** [comment]
@kevinksullivan Still overdue 6 days?! Let's take care of this!

**Comment by kbecciv** [comment]
@kevinksullivan I think you did not follow the steps:
1. Create a new WS
2. Add user to WS
3. Create room (member X is added, as he is in WS)
4. Send some message s in the room
5. Delete member X from WS
6. Send some messages in the room
I can reproduce the issue with latest build 1.2.88.0

https://user-images.githubusercontent.com/93399543/226662452-4b54108e-48fa-472d-a114-b617a8d8c539.mp4

**Comment by kevinksullivan** [comment]
Ah, thanks for breaking that down and reproducing again @kbecciv . Going to export!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~019fa2485ed55ee079

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alex-mechler (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MitchExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @joelbettner (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by alex-mechler** [comment]
@kevinksullivan you only need to add the `External` label. Adding `External` and `Engineering` can lead to 2 CMEs being assigned, like it did here 😅. going to unassign myself since @joelbettner was the one picked for the `External` label

**Comment by MelvinBot** [comment]
@joelbettner @kevinksullivan @rushatgabhane @kbecciv this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
📣 @spantheslayer! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by spantheslayer** [comment]
Contributor details
Your Expensify account email: supriyosarkar282@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~017416d65c91a9dfa3

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by spantheslayer** [comment]
@rushatgabhane thank you!! I can now access new room function.

**Comment by spantheslayer** [comment]
I have checked the mentioned Version Number: 1.2.80.0 & the latest build. Could not reproduce the issue, seems it is working as expected. Trying with version number: 1.2.88.0 as you mentioned above. Will keep posted :)

**Comment by MelvinBot** [comment]
@joelbettner, @kevinksullivan, @rushatgabhane, @kbecciv Huh... This is 4 days overdue. Who can take care of this?

**Comment by spantheslayer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Even after removing a user from a workspace, user can still send messages to any rooms related to the workspace user was a member earlier.

### What is the root cause of that problem?
In `src/libs/actions/Policy.js` , line no. 212, `pendingActions` method is deleting the user but it is not updating. 
<img width="944" alt="image" src="https://user-images.githubusercontent.com/61844611/228553989-09fd1885-dcb0-4acb-825a-5c1fdd519a6f.png">


### What changes do you think we should make in order to solve the problem?
If we change `pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE` to
`pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE` , It is updating the user side as well and user can't send the message anymore. Although the room is visible to the user side but upon sending a message user gets below error 
<img width="422" alt="image" src="https://user-images.githubusercontent.com/61844611/228555466-ef4e053a-e035-4c9b-a62e-8769cc0faccd.png">

### Working flow 
![Mar-29-2023 19-18-05](https://user-images.githubusercontent.com/61844611/228559285-68eded13-d774-4bf4-8f85-e2962c553472.gif)



### What alternative solutions did you explore? (Optional)
I have tried to update the satet reportID but that did not work. Tried to replicate the flow of how the announce room get detached everytime a user is removed from a workspace but that did not really work as I could really find the root of this flow.

**Comment by spantheslayer** [comment]
Hey @joelbettner  thanks for pointing out exactly what needed to be done. Can I do the changes and make or update the proposal?

**Comment by MelvinBot** [comment]
@joelbettner @kevinksullivan @rushatgabhane @kbecciv this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by rushatgabhane** [comment]
@joelbettner exactly my thought! I agree with all 3 changes.
There is one more case that we need to handle - when the workspace itself is deleted, we have the same bug and we should do all those 3 changes for this case as well.

Maybe you can start working on the backend changes while we land on a proposal here.

**Comment by rushatgabhane** [comment]
> Make front end changes that hide the workspace from the removed user's UI

One question - 

What exactly do we mean by removing workspace from UI? 

1. How to handle LHN
i. We could remove it completely from LHN - bad idea because you can't read previous conv.
ii. We could push the room to the bottom of chat list in LHN - this is also kinda weird because you suddenly cannot find your room. Is this acceptable? @kevinksullivan @joelbettner 
iii. We could do nothing in LHN, but add a message to the chat from backend - "User X has been removed form this room".
iv. ....?

2. How to handle room in search
i. I think we should keep the room here. Do nothing

4. Are there other places where you can see the room from? Maybe accessing the link directly

**Comment by rushatgabhane** [comment]
@spantheslayer yes please! You're welcome to drop in any thoughts, ideas, questions or an updated proposal.

**Comment by MelvinBot** [comment]
@joelbettner, @kevinksullivan, @rushatgabhane, @kbecciv Eep! 4 days overdue now. Issues have feelings too...

**Comment by kevinksullivan** [comment]
Hey @rushatgabhane ! What's the latest on this one?

**Comment by MelvinBot** [comment]
@joelbettner @kevinksullivan @rushatgabhane @kbecciv this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @rushatgabhane is eligible for the Internal assigner, not assigning anyone new.

**Comment by rushatgabhane** [comment]
on it

**Comment by rushatgabhane** [comment]
@kevinksullivan waiting on proposals and backend changes

**Comment by MelvinBot** [comment]
@joelbettner, @kevinksullivan, @rushatgabhane, @kbecciv Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@joelbettner, @kevinksullivan, @rushatgabhane, @kbecciv Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@joelbettner, @kevinksullivan, @rushatgabhane, @kbecciv Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by joelbettner** [comment]
> What exactly do we mean by removing workspace from UI?
>
> How to handle LHN
>
> We could remove it completely from LHN - bad idea because you can't read previous conv.
> We could push the room to the bottom of chat list in LHN - this is also kinda weird because you suddenly cannot find your room. Is this acceptable? @kevinksullivan @joelbettner
We could do nothing in LHN, but add a message to the chat from backend - "User X has been removed form this room".
....?
How to handle room in search
> 
> I think we should keep the room here. Do nothing
> Are there other places where you can see the room from? Maybe accessing the link directly

@rushatgabhane 

In LHN I personally think we should remove the workspace entirely if the user is removed from it. If they need to read pervious convos they can try to search it or get re-added to the workspace.

I think it should still show up in the search. So, I agree with your "do nothing" assessment.

@kevinksullivan your thoughts on this^?

**Comment by kevinksullivan** [comment]
Hm I think I agree with @rushatgabhane for now, that the chat should show and be searchable. If we want to discuss making changes beyond that they should be brought to the wider team.

**Comment by MelvinBot** [comment]
@joelbettner, @kevinksullivan, @rushatgabhane, @kbecciv Eep! 4 days overdue now. Issues have feelings too...

**Comment by kevinksullivan** [comment]
What's the latest on this one @joelbettner ?

**Comment by joelbettner** [comment]
@kevinksullivan I do not have the bandwidth to actually make these backend changes. I'm not sure what the protocol is in this situation...

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stitesExpensify (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by kevinksullivan** [comment]
Re-adding the engineering label to pick up a new internal engineer.

**Comment by stitesExpensify** [comment]
I'm going to re-assign as well. I am focused on threads today and am about to go OOO for a week.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @luacmartins (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by michaelhaxhiu** [comment]
We didn't get an assignment, so I triggered again.

**Comment by michaelhaxhiu** [comment]
PS @joelbettner and @stitesExpensify for future reference, if you get assigned a GH like this as `internal` and can't take it, can you please post in slack to find a volunteer next time? That ensures that this will land into the hands of someone who can certainly push it forward.

**Comment by luacmartins** [comment]
Hey! I just got assigned this and my plate is really full atm with Manual Requests. I posted [here](https://expensify.slack.com/archives/C03TQ48KC/p1683262294553409) to look for another volunteer to take this on, but maybe we could also re-add the engineering label?

**Comment by michaelhaxhiu** [comment]
Good question! Given that you don't have room on your plate either (and nor do the prior 2 assigned engineers), I think we need to stop using the random assignment of the `engineering` label because it's causing a delay. My recommendation is to assign the original engineer and that person should be responsible for completing the GH or finding a replacement.

Basically, we should not be unassigning GHs and re-assigning the label. Once it's assigned to to an engineer, it's their duty to ensure it's completed by them or someone else.

**Comment by flodnv** [comment]
Taking over as I am assigned to a duplicate internally https://github.com/Expensify/Expensify/issues/280702

This issue also exists when demoting someone from admin to employee, they can still chat in #admins. I intend to fix both in this issue.

**Comment by flodnv** [comment]
I reproduced, I started looking into the fix but had some questions about expected behavior: https://expensify.slack.com/archives/C04QEB4MJEQ/p1683832301167299

**Comment by flodnv** [comment]
I sent a PR to send pusher updates, but I still need to:
- Understand and investigate what happens with public and private rooms 
- Followup with another PR that ensure the member gets the chat removed from the LHN if they are offline when it happens, which is related to other bugs I'm assigned to

**Comment by flodnv** [comment]
PRs are in final review.

**Comment by flodnv** [comment]
WIP POC PR sent for the offline case.

**Comment by flodnv** [comment]
I posted a Problem/Solution for feedback here: https://expensify.slack.com/archives/C01GTK53T8Q/p1684863673526879

**Comment by flodnv** [comment]
I have been working on other room access bugs, I'll get to this at some point.

**Comment by michaelhaxhiu** [comment]
I'm reassigning Kevin, because the latest process is to always have a BZ assigned to a `Bug` regardless of Internal or External.

**Comment by kevinksullivan** [comment]
Ok, should we take this off `Daily` @flodnv ? Is that the right priority level for this?

**Comment by flodnv** [comment]
It can be weekly, yeah

**Comment by flodnv** [comment]
Can it though? https://expensify.slack.com/archives/C03U7DCU4/p1685965481640849

**Comment by michaelhaxhiu** [comment]
All bugs should be daily unless they are on hold.

**Comment by flodnv** [comment]
Hmmm strange, I was convinced this bug still existed, but after trying several times I am unsuccessful 😕 

@kevinksullivan @kbecciv is either of you able to reproduce?

**Comment by kevinksullivan** [comment]
No, I cannot reproduce it. The room was immediately hidden after user X was removed.

<img width="1369" alt="image" src="https://github.com/Expensify/App/assets/22508290/51f0193b-2aa9-4958-b335-8248e50ab5dd">

**Comment by kevinksullivan** [comment]
Dissussing here

https://expensify.slack.com/archives/C049HHMV9SM/p1686066293110289

**Comment by kevinksullivan** [comment]
Confirmed the bug is no longer reproducible!
